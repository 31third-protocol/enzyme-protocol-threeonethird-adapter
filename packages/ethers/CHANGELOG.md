# @enzymefinance/ethers

## 4.0.10

### Patch Changes

- 1f61b9af0: Release 4.0.10

## 4.0.9

### Patch Changes

- f82ac29f9: Latest protocol version

## 4.0.8

### Patch Changes

- c06baeab1: Add eth_sign for Gnosis Safe

## 4.0.7

### Patch Changes

- 8bb3e9ff6: Publish packages

## 4.0.5

### Patch Changes

- 8fe3d4a57: Publishing latest release

## 4.0.4

### Patch Changes

- 84c2e211e: Lock bn.js dependency version
- c030c92cd: Aave claimRewardsToSelf selector

## 4.0.3

### Patch Changes

- e8d978204: Upgrading ethers.js

## 4.0.2

### Patch Changes

- b4fcf8bc: Update packages after audit feedback and fixes.

## 4.0.1

### Patch Changes

- 2b54b39f: Add devtools to monorepo
