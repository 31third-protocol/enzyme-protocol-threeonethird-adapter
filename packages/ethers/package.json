{
  "name": "@enzymefinance/ethers",
  "version": "4.0.4",
  "description": "Technology Regulated and Operated Investment Funds",
  "author": "Enzyme Council DAO <council@enzyme.finance>",
  "license": "MIT",
  "homepage": "https://enzyme.finance",
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/avantgardefinance/ethereum-devtools.git"
  },
  "bugs": {
    "url": "https://github.com/avantgardefinance/ethereum-devtools/issues"
  },
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.js"
    }
  },
  "scripts": {
    "build": "tsup",
    "clean": "rimraf .turbo .eslintcache node_modules dist",
    "lint": "eslint \"src/**/*.{js,ts}\" --cache",
    "format": "yarn lint --fix",
    "typecheck": "tsc --noEmit"
  },
  "dependencies": {
    "ethers": "^5.6.8"
  },
  "devDependencies": {
    "@types/jest": "^27.4.0",
    "jest": "^27.4.7",
    "tsup": "^5.11.13",
    "typescript": "^4.5.5"
  },
  "tsup": {
    "entry": [
      "src/index.ts"
    ],
    "format": [
      "esm",
      "cjs"
    ],
    "dts": true,
    "splitting": true,
    "clean": true
  },
  "eslintConfig": {
    "extends": "@enzymefinance/eslint-config",
    "rules": {
      "@typescript-eslint/no-shadow": "off"
    },
    "parserOptions": {
      "project": [
        "./tsconfig.json"
      ]
    }
  }
}
