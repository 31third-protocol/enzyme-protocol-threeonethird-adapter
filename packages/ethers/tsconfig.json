{
  "extends": "../../tsconfig.json",
  "include": ["src"],
  "files": [],
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@enzymefinance/hardhat/types": ["../hardhat/src/types.ts"]
    }
  }
}
