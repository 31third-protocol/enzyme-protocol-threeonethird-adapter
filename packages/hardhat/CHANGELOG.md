# @enzymefinance/hardhat

## 4.0.1

### Patch Changes

- 2b54b39f: Add devtools to monorepo
- Updated dependencies [2b54b39f]
  - @enzymefinance/codegen@4.0.1
  - @enzymefinance/coverage@4.0.1
  - @enzymefinance/ethers@4.0.1
