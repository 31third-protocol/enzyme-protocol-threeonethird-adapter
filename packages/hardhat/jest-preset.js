module.exports = {
  setupFilesAfterEnv: [require.resolve('@enzymefinance/hardhat/jest/config/setup')],
  testEnvironment: require.resolve('@enzymefinance/hardhat/jest/config/environment'),
  testTimeout: 240000,
};
