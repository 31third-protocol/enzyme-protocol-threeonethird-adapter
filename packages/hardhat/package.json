{
  "name": "@enzymefinance/hardhat",
  "version": "4.0.5",
  "private": true,
  "main": "./dist/index.js",
  "exports": {
    ".": {
      "require": "./dist/index.js"
    },
    "./types": {
      "require": "./dist/types.js"
    },
    "./plugin": {
      "require": "./dist/plugin/index.js"
    },
    "./jest/config/setup": {
      "require": "./dist/jest/config/setup.js"
    },
    "./jest/config/environment": {
      "require": "./dist/jest/config/environment.js"
    },
    "./jest-preset": {
      "require": "./jest-preset.js"
    }
  },
  "scripts": {
    "build": "tsup",
    "lint": "eslint \"src/**/*.{js,ts}\" --cache",
    "format": "pnpm lint --fix",
    "typecheck": "tsc --noEmit"
  },
  "dependencies": {
    "@enzymefinance/codegen": "workspace:4.0.5",
    "@enzymefinance/ethers": "workspace:4.0.5",
    "@jest/environment": "^28.1.3",
    "@nomiclabs/hardhat-ethers": "2.0.6",
    "@nomiclabs/hardhat-etherscan": "3.0.3",
    "deepmerge": "^4.2.2",
    "ethers": "^5.6.8",
    "fs-extra": "^10.1.0",
    "glob": "^8.0.3",
    "hardhat": "2.9.4",
    "hardhat-contract-sizer": "^2.5.1",
    "hardhat-deploy": "^0.11.4",
    "jest-diff": "^28.1.3",
    "jest-environment-node": "^28.1.3",
    "jest-matcher-utils": "^28.1.3",
    "jest-snapshot": "^28.1.3",
    "ts-node": "^10.7.0"
  },
  "devDependencies": {
    "@types/fs-extra": "^9.0.13",
    "@types/jest": "^28.1.6",
    "jest": "^28.1.3",
    "tsup": "^6.2.1",
    "typescript": "^4.5.5"
  },
  "tsup": {
    "entry": [
      "src/index.ts",
      "src/types.ts",
      "src/jest/config/*.ts",
      "src/plugin/index.ts"
    ],
    "format": [
      "cjs"
    ],
    "dts": false,
    "splitting": true,
    "clean": true,
    "silent": true
  },
  "eslintConfig": {
    "extends": "@enzymefinance/eslint-config",
    "rules": {
      "@typescript-eslint/no-shadow": "off"
    },
    "parserOptions": {
      "project": [
        "./tsconfig.json"
      ]
    }
  }
}
