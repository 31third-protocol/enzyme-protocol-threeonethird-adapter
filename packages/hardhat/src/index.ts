export * from './provider';
export * from './history';
export * from './snapshots';
export * from './signer';

export { setIgnoreGasMatchers } from './jest/matchers/functions/common/ignoreGasMatchers';
