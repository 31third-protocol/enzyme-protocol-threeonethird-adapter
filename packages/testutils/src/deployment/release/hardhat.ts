import {
  Dispatcher,
  VaultLib,
  FundDeployer,
  PolicyManager,
  AavePriceFeed,
  AlphaHomoraV1PriceFeed,
  ChaiPriceFeed,
  ChainlinkPriceFeed,
  CompoundPriceFeed,
  CurvePriceFeed,
  LidoStethPriceFeed,
  StakehoundEthPriceFeed,
  SynthetixPriceFeed,
  WdgldPriceFeed,
  AggregatedDerivativePriceFeed,
  ValueInterpreter,
  UniswapV2PoolPriceFeed,
  IntegrationManager,
  CurveLiquidityStethAdapter,
  AaveAdapter,
  AlphaHomoraV1Adapter,
  ParaSwapAdapter,
  SynthetixAdapter,
  ZeroExV2Adapter,
  CompoundAdapter,
  UniswapV2Adapter,
  TrackedAssetsAdapter,
  ChaiAdapter,
  CurveExchangeAdapter,
  KyberAdapter,
  FeeManager,
  ComptrollerLib,
  EntranceRateBurnFee,
  EntranceRateDirectFee,
  ManagementFee,
  PerformanceFee,
  AuthUserExecutedSharesRequestorLib,
  AuthUserExecutedSharesRequestorFactory,
  FundActionsWrapper,
  AdapterBlacklist,
  AdapterWhitelist,
  AssetBlacklist,
  AssetWhitelist,
  BuySharesCallerWhitelist,
  MaxConcentration,
  MinMaxInvestment,
  InvestorWhitelist,
  GuaranteedRedemption,
} from '@enzymefinance/protocol';

import { DeploymentConfig } from '../../../../../deploy/utils/config';

export type ForkDeployment = ReturnType<typeof loadForkDeployment> extends Promise<infer T> ? T : never;

export async function loadForkDeployment() {
  const output = await hre.deployments.fixture();
  const named = await hre.getNamedAccounts();
  const unnamed = await hre.getUnnamedAccounts();

  const deployer = await provider.getSignerWithAddress(named.deployer);
  const accounts = await Promise.all(unnamed.map((account) => provider.getSignerWithAddress(account)));
  const config = output['Config'].linkedData as DeploymentConfig;

  // prettier-ignore
  const deployment = {
    Dispatcher: new Dispatcher(output['Dispatcher'].address, deployer),
    VaultLib: new VaultLib(output['VaultLib'].address, deployer),
    FundDeployer: new FundDeployer(output['FundDeployer'].address, deployer),
    PolicyManager: new PolicyManager(output['PolicyManager'].address, deployer),
    AavePriceFeed: new AavePriceFeed(output['AavePriceFeed'].address, deployer),
    AlphaHomoraV1PriceFeed: new AlphaHomoraV1PriceFeed(output['AlphaHomoraV1PriceFeed'].address, deployer),
    ChaiPriceFeed: new ChaiPriceFeed(output['ChaiPriceFeed'].address, deployer),
    ChainlinkPriceFeed: new ChainlinkPriceFeed(output['ChainlinkPriceFeed'].address, deployer),
    CompoundPriceFeed: new CompoundPriceFeed(output['CompoundPriceFeed'].address, deployer),
    CurvePriceFeed: new CurvePriceFeed(output['CurvePriceFeed'].address, deployer),
    LidoStethPriceFeed: new LidoStethPriceFeed(output['LidoStethPriceFeed'].address, deployer),
    SynthetixPriceFeed: new SynthetixPriceFeed(output['SynthetixPriceFeed'].address, deployer),
    StakehoundEthPriceFeed: new StakehoundEthPriceFeed(output['StakehoundEthPriceFeed'].address, deployer),
    WdgldPriceFeed: new WdgldPriceFeed(output['WdgldPriceFeed'].address, deployer),
    AggregatedDerivativePriceFeed: new AggregatedDerivativePriceFeed(output['AggregatedDerivativePriceFeed'].address, deployer),
    ValueInterpreter: new ValueInterpreter(output['ValueInterpreter'].address, deployer),
    UniswapV2PoolPriceFeed: new UniswapV2PoolPriceFeed(output['UniswapV2PoolPriceFeed'].address, deployer),
    IntegrationManager: new IntegrationManager(output['IntegrationManager'].address, deployer),
    CurveLiquidityStethAdapter: new CurveLiquidityStethAdapter(output['CurveLiquidityStethAdapter'].address, deployer),
    AaveAdapter: new AaveAdapter(output['AaveAdapter'].address, deployer),
    AlphaHomoraV1Adapter: new AlphaHomoraV1Adapter(output['AlphaHomoraV1Adapter'].address, deployer),
    ParaSwapAdapter: new ParaSwapAdapter(output['ParaSwapAdapter'].address, deployer),
    SynthetixAdapter: new SynthetixAdapter(output['SynthetixAdapter'].address, deployer),
    ZeroExV2Adapter: new ZeroExV2Adapter(output['ZeroExV2Adapter'].address, deployer),
    CompoundAdapter: new CompoundAdapter(output['CompoundAdapter'].address, deployer),
    UniswapV2Adapter: new UniswapV2Adapter(output['UniswapV2Adapter'].address, deployer),
    TrackedAssetsAdapter: new TrackedAssetsAdapter(output['TrackedAssetsAdapter'].address, deployer),
    ChaiAdapter: new ChaiAdapter(output['ChaiAdapter'].address, deployer),
    CurveExchangeAdapter: new CurveExchangeAdapter(output['CurveExchangeAdapter'].address, deployer),
    KyberAdapter: new KyberAdapter(output['KyberAdapter'].address, deployer),
    FeeManager: new FeeManager(output['FeeManager'].address, deployer),
    ComptrollerLib: new ComptrollerLib(output['ComptrollerLib'].address, deployer),
    EntranceRateBurnFee: new EntranceRateBurnFee(output['EntranceRateBurnFee'].address, deployer),
    EntranceRateDirectFee: new EntranceRateDirectFee(output['EntranceRateDirectFee'].address, deployer),
    ManagementFee: new ManagementFee(output['ManagementFee'].address, deployer),
    PerformanceFee: new PerformanceFee(output['PerformanceFee'].address, deployer),
    AuthUserExecutedSharesRequestorLib: new AuthUserExecutedSharesRequestorLib(output['AuthUserExecutedSharesRequestorLib'].address, deployer),
    AuthUserExecutedSharesRequestorFactory: new AuthUserExecutedSharesRequestorFactory(output['AuthUserExecutedSharesRequestorFactory'].address, deployer),
    FundActionsWrapper: new FundActionsWrapper(output['FundActionsWrapper'].address, deployer),
    AdapterBlacklist: new AdapterBlacklist(output['AdapterBlacklist'].address, deployer),
    AdapterWhitelist: new AdapterWhitelist(output['AdapterWhitelist'].address, deployer),
    AssetBlacklist: new AssetBlacklist(output['AssetBlacklist'].address, deployer),
    AssetWhitelist: new AssetWhitelist(output['AssetWhitelist'].address, deployer),
    BuySharesCallerWhitelist: new BuySharesCallerWhitelist(output['BuySharesCallerWhitelist'].address, deployer),
    MaxConcentration: new MaxConcentration(output['MaxConcentration'].address, deployer),
    MinMaxInvestment: new MinMaxInvestment(output['MinMaxInvestment'].address, deployer),
    InvestorWhitelist: new InvestorWhitelist(output['InvestorWhitelist'].address, deployer),
    GuaranteedRedemption: new GuaranteedRedemption(output['GuaranteedRedemption'].address, deployer),
  } as const;

  return {
    deployer,
    deployment,
    config,
    accounts,
  } as const;
}
