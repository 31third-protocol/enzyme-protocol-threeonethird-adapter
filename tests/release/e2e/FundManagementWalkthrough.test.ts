import { extractEvent, randomAddress } from '@enzymefinance/ethers';
import { SignerWithAddress } from '@enzymefinance/hardhat';
import {
  ComptrollerLib,
  convertRateToScaledPerSecondRate,
  entranceRateFeeConfigArgs,
  feeManagerConfigArgs,
  investorWhitelistArgs,
  managementFeeConfigArgs,
  performanceFeeConfigArgs,
  StandardToken,
  VaultLib,
} from '@enzymefinance/protocol';
import {
  addTrackedAssetsToVault,
  buyShares,
  createNewFund,
  deployProtocolFixture,
  ProtocolDeployment,
  redeemSharesInKind,
  uniswapV2TakeOrder,
} from '@enzymefinance/testutils';
import { BigNumber, BigNumberish, utils } from 'ethers';

const expectedGasCosts = {
  'buy shares: denomination asset only: first investment': {
    usdc: 381000,
    weth: 359000,
  },
  'buy shares: denomination asset only: second investment': {
    usdc: 334000,
    weth: 319000,
  },
  'buy shares: max assets': {
    usdc: 1429000,
    weth: 1356000,
  },
  'calc gav: 20 assets': {
    usdc: 1110000,
    weth: 1072000,
  },
  'calc gav: denomination asset only': {
    usdc: 84400,
    weth: 76000,
  },

  'create fund': {
    usdc: 933000,
    weth: 922000,
  },

  'redeem all shares: max assets: all remaining': {
    usdc: 2500000,
    weth: 2490000,
  },

  'redeem partial shares: max assets': {
    usdc: 2590000,
    weth: 2575000,
  },

  'trade on Uniswap: max assets': {
    usdc: 269000,
    weth: 245000,
  },
} as const;

describe.each([['weth' as const], ['usdc' as const]])(
  'Walkthrough for %s as denomination asset',
  (denominationAssetId) => {
    let fork: ProtocolDeployment;
    let manager: SignerWithAddress;
    let investor: SignerWithAddress;
    let anotherInvestor: SignerWithAddress;

    let comptrollerProxy: ComptrollerLib;
    let vaultProxy: VaultLib;
    let denominationAsset: StandardToken;
    let denominationAssetDecimals: BigNumberish;

    beforeAll(async () => {
      fork = await deployProtocolFixture();

      manager = fork.accounts[1];
      investor = fork.accounts[2];
      anotherInvestor = fork.accounts[3];

      denominationAsset =
        denominationAssetId === 'weth'
          ? new StandardToken(fork.config.weth, whales.weth)
          : new StandardToken(fork.config.primitives[denominationAssetId], whales[denominationAssetId]);
      denominationAssetDecimals = await denominationAsset.decimals();

      // Seed investors with denomination asset
      const denominationAssetSeedAmount = utils.parseUnits('100', await denominationAsset.decimals());
      await denominationAsset.transfer(investor, denominationAssetSeedAmount);
      await denominationAsset.transfer(anotherInvestor, denominationAssetSeedAmount);
    });

    it('creates a new fund', async () => {
      // fees
      const rate = utils.parseEther('0.01');
      const scaledPerSecondRate = convertRateToScaledPerSecondRate(rate);

      const managementFeeSettings = managementFeeConfigArgs(scaledPerSecondRate);
      const performanceFeeSettings = performanceFeeConfigArgs({
        rate: utils.parseEther('0.1'),
        period: 365 * 24 * 60 * 60,
      });
      const entranceRateFeeSettings = entranceRateFeeConfigArgs(utils.parseEther('0.05'));

      const feeManagerConfig = feeManagerConfigArgs({
        fees: [fork.deployment.managementFee, fork.deployment.performanceFee, fork.deployment.entranceRateBurnFee],
        settings: [managementFeeSettings, performanceFeeSettings, entranceRateFeeSettings],
      });

      // TODO: add policies

      const createFundTx = await createNewFund({
        signer: manager,
        fundDeployer: fork.deployment.fundDeployer,
        fundOwner: manager,
        denominationAsset,
        feeManagerConfig,
      });

      comptrollerProxy = createFundTx.comptrollerProxy;
      vaultProxy = createFundTx.vaultProxy;

      expect(createFundTx.receipt).toCostLessThan(expectedGasCosts['create fund'][denominationAssetId]);
    });

    it('enables the InvestorWhitelist policy for the fund', async () => {
      const enabled = await fork.deployment.policyManager
        .connect(manager)
        .enablePolicyForFund.args(
          comptrollerProxy.address,
          fork.deployment.investorWhitelist,
          investorWhitelistArgs({
            investorsToAdd: [randomAddress(), randomAddress(), investor.address],
          }),
        )
        .send();

      expect(enabled).toBeReceipt();
    });

    it('buys shares of a fund', async () => {
      const buySharesTx = await buyShares({
        comptrollerProxy,
        buyer: investor,
        denominationAsset,
      });

      const rate = utils.parseEther('0.05');
      const rateDivisor = utils.parseEther('1');
      const expectedFee = utils.parseUnits('1', denominationAssetDecimals).mul(rate).div(rateDivisor.add(rate));

      expect(await vaultProxy.balanceOf(investor)).toBeGteBigNumber(
        utils.parseUnits('1', denominationAssetDecimals).sub(expectedFee),
      );

      expect(buySharesTx).toCostLessThan(
        expectedGasCosts['buy shares: denomination asset only: first investment'][denominationAssetId],
      );
    });

    it('buys more shares of a fund', async () => {
      const previousBalance = await vaultProxy.balanceOf(investor);

      const minSharesAmount = utils.parseUnits('0.00001', denominationAssetDecimals);
      const buySharesTx = await buyShares({
        comptrollerProxy,
        buyer: investor,
        denominationAsset,
      });

      expect(await vaultProxy.balanceOf(investor)).toBeGteBigNumber(minSharesAmount.add(previousBalance));

      expect(buySharesTx).toCostLessThan(
        expectedGasCosts['buy shares: denomination asset only: second investment'][denominationAssetId],
      );
    });

    it('calculates the GAV of the fund with only the denomination asset', async () => {
      const calcGavTx = await comptrollerProxy.calcGav(true);

      expect(calcGavTx).toCostLessThan(expectedGasCosts['calc gav: denomination asset only'][denominationAssetId]);
    });

    it('seeds the fund with all more assets', async () => {
      const assets = [
        new StandardToken(fork.config.primitives.bat, whales.bat),
        new StandardToken(fork.config.primitives.bnb, whales.bnb),
        new StandardToken(fork.config.primitives.bnt, whales.bnt),
        new StandardToken(fork.config.primitives.comp, whales.comp),
        new StandardToken(fork.config.primitives.dai, whales.dai),
        new StandardToken(fork.config.primitives.link, whales.link),
        new StandardToken(fork.config.primitives.mana, whales.mana),
        new StandardToken(fork.config.primitives.mln, whales.mln),
        new StandardToken(fork.config.primitives.ren, whales.ren),
        new StandardToken(fork.config.primitives.rep, whales.rep),
        new StandardToken(fork.config.primitives.susd, whales.susd),
        new StandardToken(fork.config.primitives.uni, whales.uni),
        new StandardToken(fork.config.primitives.usdt, whales.usdt),
        new StandardToken(fork.config.primitives.zrx, whales.zrx),
      ];

      await addTrackedAssetsToVault({
        signer: manager,
        comptrollerProxy,
        integrationManager: fork.deployment.integrationManager,
        assets,
      });

      // Use this loop instead of addNewAssetsToFund() to make debugging easier
      // when a whale changes.
      for (const asset of assets) {
        const decimals = await asset.decimals();
        const transferAmount = utils.parseUnits('1', decimals);
        await asset.transfer.args(vaultProxy, transferAmount).send();

        const balance = await asset.balanceOf(vaultProxy);
        expect(balance).toBeGteBigNumber(transferAmount);
      }
    });

    it('seeds the fund with cTokens', async () => {
      const compoundAssets = [
        new StandardToken(fork.config.compound.ctokens.ccomp, whales.ccomp),
        new StandardToken(fork.config.compound.ctokens.cdai, whales.cdai),
        new StandardToken(fork.config.compound.ceth, whales.ceth),
        new StandardToken(fork.config.compound.ctokens.cusdc, whales.cusdc),
        new StandardToken(fork.config.compound.ctokens.cuni, whales.cuni),
      ];

      await addTrackedAssetsToVault({
        signer: manager,
        comptrollerProxy,
        integrationManager: fork.deployment.integrationManager,
        assets: compoundAssets,
      });

      // Use this loop instead of addNewAssetsToFund() to make debugging easier
      // when a whale changes.
      for (const asset of compoundAssets) {
        const decimals = await asset.decimals();
        const transferAmount = utils.parseUnits('1', decimals);

        await asset.transfer.args(vaultProxy, transferAmount).send();

        const balance = await asset.balanceOf(vaultProxy);
        expect(balance).toBeGteBigNumber(transferAmount);
      }
    });

    it('calculates the GAV of the fund with 20 assets', async () => {
      expect((await vaultProxy.getTrackedAssets()).length).toBe(20);

      const calcGavTx = await comptrollerProxy.calcGav(true);

      expect(calcGavTx).toCostLessThan(expectedGasCosts['calc gav: 20 assets'][denominationAssetId]);
    });

    it('trades on Uniswap', async () => {
      const receipt = await uniswapV2TakeOrder({
        comptrollerProxy,
        vaultProxy,
        integrationManager: fork.deployment.integrationManager,
        fundOwner: manager,
        uniswapV2Adapter: fork.deployment.uniswapV2Adapter,
        path: [denominationAsset, new StandardToken(fork.config.primitives.dai, provider)],
        outgoingAssetAmount: utils.parseUnits('0.1', denominationAssetDecimals),
        minIncomingAssetAmount: BigNumber.from(1),
      });

      expect(receipt).toCostLessThan(expectedGasCosts['trade on Uniswap: max assets'][denominationAssetId]);
    });

    it("sends an asset amount to the fund's vault", async () => {
      const gavBefore = await comptrollerProxy.calcGav.args(true).call();
      const grossShareValueBefore = await comptrollerProxy.calcGrossShareValue.call();

      const dai = new StandardToken(fork.config.primitives.dai, whales.dai);
      const amount = utils.parseEther('1');

      await dai.transfer(vaultProxy, amount);

      const gavAfter = await comptrollerProxy.calcGav.args(true).call();
      const grossShareValueAfter = await comptrollerProxy.calcGrossShareValue.call();

      expect(gavAfter.gav_).toBeGtBigNumber(gavBefore.gav_);
      expect(grossShareValueAfter.grossShareValue_).toBeGtBigNumber(grossShareValueBefore.grossShareValue_);
    });

    it('redeems some shares of the investor (without fees failure)', async () => {
      const balance = await vaultProxy.balanceOf(investor);
      const redeemQuantity = balance.div(2);

      const redeemed = await redeemSharesInKind({
        comptrollerProxy,
        signer: investor,
        quantity: redeemQuantity,
      });

      const failureEvents = extractEvent(redeemed, 'PreRedeemSharesHookFailed');
      expect(failureEvents.length).toBe(0);

      expect(await vaultProxy.balanceOf(investor)).toEqBigNumber(balance.sub(redeemQuantity));

      expect(redeemed).toCostLessThan(expectedGasCosts['redeem partial shares: max assets'][denominationAssetId]);
    });

    it("sends an asset amount to the fund's vault again", async () => {
      const gavBefore = await comptrollerProxy.calcGav.args(true).call();
      const grossShareValueBefore = await comptrollerProxy.calcGrossShareValue.call();

      const zrx = new StandardToken(fork.config.primitives.zrx, whales.zrx);
      const amount = utils.parseEther('1');

      await zrx.transfer(vaultProxy, amount);

      const gavAfter = await comptrollerProxy.calcGav.args(true).call();
      const grossShareValueAfter = await comptrollerProxy.calcGrossShareValue.call();

      expect(gavAfter.gav_).toBeGtBigNumber(gavBefore.gav_);
      expect(grossShareValueAfter.grossShareValue_).toBeGtBigNumber(grossShareValueBefore.grossShareValue_);
    });

    it('changes the InvestorWhitelist', async () => {
      await fork.deployment.policyManager
        .connect(manager)
        .updatePolicySettingsForFund.args(
          comptrollerProxy.address,
          fork.deployment.investorWhitelist,
          investorWhitelistArgs({
            investorsToAdd: [anotherInvestor],
            investorsToRemove: [investor],
          }),
        )
        .send();
    });

    it('buy shares: max assets', async () => {
      const investmentAmount = utils.parseUnits('1', denominationAssetDecimals);

      const grossShareValue = await comptrollerProxy.calcGrossShareValue.call();
      const minSharesQuantity = investmentAmount
        .mul(utils.parseEther('1'))
        .div(grossShareValue.grossShareValue_)
        .mul(95) // deduct 5% for safety
        .div(100);

      const buySharesTx = await buyShares({
        comptrollerProxy,
        buyer: anotherInvestor,
        denominationAsset,
        investmentAmount,
        minSharesQuantity,
      });

      expect(await vaultProxy.balanceOf(anotherInvestor)).toBeGteBigNumber(minSharesQuantity);

      expect(buySharesTx).toCostLessThan(expectedGasCosts['buy shares: max assets'][denominationAssetId]);
    });

    it('redeems all remaining shares of the first investor (without fees failure)', async () => {
      const redeemed = await redeemSharesInKind({
        comptrollerProxy,
        signer: investor,
      });

      const failureEvents = extractEvent(redeemed, 'PreRedeemSharesHookFailed');
      expect(failureEvents.length).toBe(0);

      expect(await vaultProxy.balanceOf(investor)).toEqBigNumber(utils.parseEther('0'));
    });

    it('redeems all remaining shares of the other investor (without fees failure)', async () => {
      const redeemed = await redeemSharesInKind({
        comptrollerProxy,
        signer: anotherInvestor,
      });

      const failureEvents = extractEvent(redeemed, 'PreRedeemSharesHookFailed');
      expect(failureEvents.length).toBe(0);

      expect(await vaultProxy.balanceOf(anotherInvestor)).toEqBigNumber(utils.parseEther('0'));

      expect(redeemed).toCostLessThan(
        expectedGasCosts['redeem all shares: max assets: all remaining'][denominationAssetId],
      );
    });
  },
);
