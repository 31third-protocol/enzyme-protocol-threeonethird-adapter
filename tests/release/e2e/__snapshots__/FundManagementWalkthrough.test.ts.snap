// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Walkthrough for usdc as denomination asset buy shares: max assets: usdc 1`] = `1214268`;

exports[`Walkthrough for usdc as denomination asset buys more shares of a fund: usdc 1`] = `368775`;

exports[`Walkthrough for usdc as denomination asset buys shares of a fund: usdc 1`] = `332009`;

exports[`Walkthrough for usdc as denomination asset calculates the GAV of the fund with 20 assets: usdc 1`] = `929243`;

exports[`Walkthrough for usdc as denomination asset calculates the GAV of the fund with only the denomination asset: usdc 1`] = `66869`;

exports[`Walkthrough for usdc as denomination asset creates a new fund: usdc 1`] = `871257`;

exports[`Walkthrough for usdc as denomination asset redeems all remaining shares of the first investor (without fees failure): usdc 1`] = `1864895`;

exports[`Walkthrough for usdc as denomination asset redeems some shares of the investor (without fees failure): usdc 1`] = `2307503`;

exports[`Walkthrough for usdc as denomination asset trades on Uniswap: usdc 1`] = `255609`;

exports[`Walkthrough for weth as denomination asset buy shares: max assets: weth 1`] = `1116385`;

exports[`Walkthrough for weth as denomination asset buys more shares of a fund: weth 1`] = `347308`;

exports[`Walkthrough for weth as denomination asset buys shares of a fund: weth 1`] = `309341`;

exports[`Walkthrough for weth as denomination asset calculates the GAV of the fund with 20 assets: weth 1`] = `843165`;

exports[`Walkthrough for weth as denomination asset calculates the GAV of the fund with only the denomination asset: weth 1`] = `57330`;

exports[`Walkthrough for weth as denomination asset creates a new fund: weth 1`] = `860740`;

exports[`Walkthrough for weth as denomination asset redeems all remaining shares of the first investor (without fees failure): weth 1`] = `1770663`;

exports[`Walkthrough for weth as denomination asset redeems some shares of the investor (without fees failure): weth 1`] = `2213089`;

exports[`Walkthrough for weth as denomination asset trades on Uniswap: weth 1`] = `229815`;
